include/gul14/to_number.h

```diff
@@ -500,7 +500,7 @@ constexpr inline optional<NumberType> to_number(gul14::string_view str) noexcept
         else if (*result > max_abs_negative_value)
             return nullopt;
 
-        return -static_cast<NumberType>(*result);
+        return static_cast<NumberType>(-(*result));
```

Comment:
The problem with the original code is that `static_cast<char>(*result)` undergoes integer promotion because of the unary `-` in front of it. Effectively this returns an `int`, which has to be implicitly cast back to `char`.

With the new code, the integer promotion happens to `(*result)`, and the unary minus should implicitly convert to an appropriate integer type (?) before we do the final `static_cast` to the return type. I think this is only valid because we have checks in the lines before that make sure that there actually is a signed integer type that can hold `-(*result)`... does that make any sense? :-/

Reply:
Maybe this?

```suggestion
        return std::negate<NumberType>()(*result));
```

Reply:
There is, of course, the ugly and maximally explicit version:
```cpp
return static_cast<NumberType>(-static_cast<NumberType>(*result));
```

Reply:
... I put the last version into the commit.

Reply:
![image](https://user-images.githubusercontent.com/16012374/225628895-46339cf7-693a-477a-aece-0f43520afc58.png)